app/models/reports/repository.rb

```diff
@@ -0,0 +1,100 @@
+module Reports
+  class Repository
+    def initialize(regions, periods:)
+      @regions = Array(regions)
+      @regions_by_id = @regions.map { |r| [r.id, r] }.to_h
+
+      @periods = if periods.is_a?(Period)
+        Range.new(periods, periods).to_a
+      else
+        periods.to_a
+      end
+    end
+
+    attr_reader :regions, :periods
+    attr_reader :regions_by_id
+    delegate :cache, :logger, to: Rails
+
+    class CacheKey
```

Comment:
This handles extracting out pieces of the cache key that we need for the queries.